Resolver gatherer

rtorrero opened the pull request:
This PR adds a resolver gatherer, it uses the filesystem to search for running sap systems and the profile data to get the virtual hostnames associated to each instance of the sap system. It will then attempt to resolve those hostnames to confirm that the OS knows about them.

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,146 @@
+	return sysNames, nil
+}
+
+func (r *SapLocalhostResolverGatherer) getProfiles() ([]map[string]string, error) {
```
I think we could simplify this by (pseudocode):
``` 
func ... getInstanceHostnames([]string, error) {
	systems, err := sapsystem.FindSystems(r.fs)

	if err != nil {
		return nil, err
	}

        for _, system := range systems {
		sid := filepath.Base(system)

                profileFiles, err: = FindProfiles(sid) {
                     hostname, err := applyRegexPatternToFindHostnameInProfileFileName...
                     hostnames = append(hostnames, hostname)
                }
        }

        return hostnames, nil

}
```

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,146 @@
+
+	for _, factReq := range factsRequests {
+		var fact entities.Fact
+		for _, profile := range profiles {
```
We iterate here over hostnames better using the `getInstanceHostnames` function

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,146 @@
+				// 	}
+				// }
+
+				factValue := &entities.FactValueMap{
```
You don't really need to create the `FactValueMap` yourself.
You can create your own internal struct with its own json annotations, and later on marshall/unmarshall/convertToFactValue as we do in many other gatherers

arbulu89 reviewed (commented):
@rtorrero Some initial comments about the generic logic of the code

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -28,6 +35,12 @@ type SapLocalhostResolverGatherer struct {
 	hr utils.HostnameResolver
 }
 
+type ReachabilityDetails struct {
```
Wouldn't be `ResorvabilityDetails` better? (what a name :laughing: )

arbulu89 commented on internal/factsengine/gatherers/gatherer.go:
```diff
@@ -10,20 +10,21 @@ type FactGatherer interface {
+		FstabGathererName:                NewDefaultFstabGatherer(),
+		GroupsGathererName:               NewDefaultGroupsGatherer(),
+		HostsFileGathererName:            NewDefaultHostsFileGatherer(),
+		SapLocalhostResolverGathererName: NewDefaultSapLocalhostResolverGatherer(),
```
Alphabetical order please

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,123 @@
+)
+
+const (
+	SapLocalhostResolverGathererName = "saplocalhost_resolver"
```
Let's decide the naming after we have the code hehe

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,123 @@
+
+func (r *SapLocalhostResolverGatherer) getInstanceHostnameDetails() (map[string]ReachabilityDetails, error) {
+	systems, err := sapsystem.FindSystems(r.fs)
+	reachabilityDetails := make(map[string]ReachabilityDetails)
```
You could initialized this after the next error checking, as it is not needed until that point

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,123 @@
+		for _, profileFile := range profileFiles {
+			result := make(map[string]string)
+			match := hostnameRegexCompiled.FindStringSubmatch(profileFile)
+			if len(match) == 0 {
```
It wouldn't be better doing `if len(match) != 4 { continue }` and using the `match[1]`, `match[2]`, etc
I know that the `SubexpNames` looks useful, but it makes the code more complex in my opinion, as it includes the need for a new map variable and a `for+if` loops

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,123 @@
+				Addresses:    []string{},
+				InstanceName: result["InstanceName"],
+			}
+			details.Addresses, _ = r.hr.LookupHost(details.Hostname)
```
I think it would be better to run this before creating the details (check the error), and compose the details as:
```
addresses, err = r.hr.LookupHost(details.Hostname)
if err != nil {
    return nil, err
}
details := eachabilityDetails{
    Hostname: match[3]
    Addresses:    addresses,
    InstanceName: match[2],
}
```

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,123 @@
+				InstanceName: result["InstanceName"],
+			}
+			details.Addresses, _ = r.hr.LookupHost(details.Hostname)
+			reachabilityDetails[result["SID"]] = details
```
I guess this is `wip` yet

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,123 @@
+
+	for _, factReq := range factsRequests {
+		var fact entities.Fact
+		factValue, _ := mapReachabilityDetailsToFactValue(rd)
```
check for errors please

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver_test.go:
```diff
@@ -0,0 +1,95 @@
+	err = appFS.MkdirAll("/usr/sap/QAS", 0644)
+	suite.NoError(err)
+
+	defaultProfileFile, _ := os.Open(helpers.GetFixturePath("gatherers/sap_profile.default"))
```
No need for this. You can create the files with `[]byte{}` as you don't care the content

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver_test.go:
```diff
@@ -0,0 +1,95 @@
+	defaultProfileContent, _ := io.ReadAll(defaultProfileFile)
+	err = afero.WriteFile(appFS, "/sapmnt/PRD/profile/DEFAULT.PFL", defaultProfileContent, 0644)
+	suite.NoError(err)
+	err = afero.WriteFile(appFS, "/sapmnt/PRD/profile/DEFAULT.1.PFL", []byte{}, 0644)
```
We don't need this now. The `FindProfiles` is already tested and we know that it excludes the backups

arbulu89 reviewed (commented):
Looking good @rtorrero 
Good progress

arbulu89 commented on pkg/utils/hostnameresolver.go:
```diff
@@ -0,0 +1,15 @@
+
+//go:generate mockery --name=HostnameResolver
+
+type HostnameResolver interface {
```
Maybe we don't need a new file for this. This interface is only used by this gatherer, so we could have it directly in the gatherer file.
We already to this with other gatherers: https://github.com/trento-project/agent/blob/main/internal/factsengine/gatherers/systemd.go#L34

And you pass the `Resolver` or `DefaultResolver` from golang net package:
https://cs.opensource.google/go/go/+/master:src/net/lookup.go;l=128

`net.DefaultResolver{}` and use it as `r.LookupHost(context.Background(), host)`

Anyway, not a big deal

CDimonaco replied:
I completely agree, and anyway I think it's not the place in package utils, If you want to put in a different file you can have it in the gatherers package, like we did with groups and users lookup for the dir-scan gatherer

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,146 @@
+		}
+
+		for _, profileFile := range profileFiles {
+			if filepath.Base(profileFile) == "DEFAULT.PFL" {
```
why `filepath.Base`? The profile file is `DEFAULT.PFL` already

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,146 @@
+
+func (r *SapLocalhostResolverGatherer) getInstanceHostnameDetails() (map[string]ResolvabilityDetails, error) {
+	systems, err := sapsystem.FindSystems(r.fs)
+	reachabilityDetails := make(map[string]ResolvabilityDetails)
```
This should be `make(map[string][]ResolvabilityDetails)`. We need a list of `ResolvabilityDetails`, as we check for each hostname.

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,146 @@
+				Addresses:    addresses,
+				InstanceName: match[2],
+			}
+			reachabilityDetails[match[1]] = details
```
```
if detailsList, ok := reachabilityDetails[match[1]]; !ok {
    reachabilityDetails[match[1]] = []ResolvabilityDetails{details}
} else {
    reachabilityDetails[match[1]] = append(reachabilityDetails[match[1]], details)
}
```
or something like that, otherwise you are overwritting all the time

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,146 @@
+	return reachabilityDetails, nil
+}
+
+func mapReachabilityDetailsToFactValue(entries map[string]ResolvabilityDetails) (entities.FactValue, error) {
```
`map[string][]ResolvabilityDetails`

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,146 @@
+func (r *SapLocalhostResolverGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	var facts []entities.Fact
+
+	rd, err := r.getInstanceHostnameDetails()
```
`rd` ? maybe better `details`

arbulu89 reviewed (commented):
More comments

CDimonaco commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+func (r *SapLocalhostResolverGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	facts := make([]entities.Fact, 0, len(factsRequests))
```
Please use a direct slice creation without make with a fixed capacity, the code looks less readable and I think we don't need such a thing in this context performance wise

CDimonaco commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,154 @@
+				Addresses:    addresses,
+				InstanceName: match[2],
+			}
+			if _, ok := resolvabilityDetails[match[1]]; !ok {
```
Could you use found, instead of ok? Ok is used more in typecasting scenarios, found is more appropriate if you are checking if you found something in the map

CDimonaco commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,154 @@
+
+// nolint:gochecknoglobals
+var (
+	subgroupMapping = map[int]string{
```
Could we squash this in the sprintf below? Wdyt? Completely up to you, We use is with len in other code, maybe a map is not the appropriate structure?

CDimonaco commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,154 @@
+			}
+
+			match := hostnameRegexCompiled.FindStringSubmatch(profileFile)
+			if len(match) != len(subgroupMapping)+1 {
```
The len thing, as commented above

CDimonaco commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,154 @@
+			}
+
+			details := ResolvabilityDetails{
+				Hostname:     match[3],
```
Could we assign the match results to a local variables so it's more readable, instead of directly accessing with index every time?

CDimonaco commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,154 @@
+		}
+
+		for _, profileFile := range profileFiles {
+			if profileFile == "DEFAULT.PFL" {
```
Could we use a constant here?

CDimonaco reviewed (changes requested):
I think the code looks good, I'm requesting changes not on the logic itself, but on style and readability

Anyway GJ

rtorrero commented:
@CDimonaco @arbulu89 thanks for the reviews, I think I've addressed everything, the only thing that I'm not 100% I did accordingly was the mocks, I didn't fully understand @arbulu89 proposal but I think I came up with a solution that addresses the main concern of not adding stuff to the utils when it's not being used elsewhere. In addition to the comments, I also:
 - Added the 'reachability' details (using ping) in the resulting fact value.
 - Improved a bit the main success test to include additional profiles and a failing ping.

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,177 @@
+
+	for _, factReq := range factsRequests {
+		var fact entities.Fact
+		factValue, err := mapReachabilityDetailsToFactValue(details)
```
We don't need to run `mapReachabilityDetailsToFactValue` within the loop. It uses always the same argument.
We should run it outside

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,177 @@
+		sid := filepath.Base(system)
+		profileFiles, err := sapsystem.FindProfiles(r.fs, sid)
+		if err != nil {
+			return nil, SapLocalhostResolverFileSystemError.Wrap(err.Error())
```
return `err` here

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,177 @@
+
+			match := hostnameRegexCompiled.FindStringSubmatch(profileFile)
+			if len(match) != regexSubgroupsCount {
+				continue
```
This `if len(match) != regexSubgroupsCount` is an exception, we should print some warning message I guess

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,177 @@
+
+			addresses, err := r.hr.LookupHost(matchedHostname)
+			if err != nil {
+				return nil, SapLocalhostResolverHostnameResolutionError.Wrap(err.Error())
```
We shouldn't error in this case. We should set an empty list of addresses.
Otherwise, a failure in 1 hostname condemns the fact gathering to fail, while some other honstnames might be resolvable

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,177 @@
+				InstanceName: matchedInstanceName,
+				Reachability: r.hp.Ping(matchedHostname),
+			}
+			if _, found := resolvabilityDetails[match[1]]; !found {
```
`match[1]` -> `matchedSID`.
And maybe a new line before the `if` to improve readability

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver_test.go:
```diff
@@ -0,0 +1,154 @@
+	"github.com/spf13/afero"
+	"github.com/stretchr/testify/suite"
+	"github.com/trento-project/agent/internal/factsengine/gatherers"
+	mocks "github.com/trento-project/agent/internal/factsengine/gatherers/mocks"
```
No need for this `mocks` alias, you don't change name after all

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver_test.go:
```diff
@@ -0,0 +1,154 @@
+	err := appFS.MkdirAll("/usr/sap/QAS", 0644)
+	suite.NoError(err)
+
+	ascsProfileFile, _ := os.Open(helpers.GetFixturePath("gatherers/sap_profile.ascs"))
```
I don't think we need to get the content of a file. We don't use it.
```
err = afero.WriteFile(appFS, "/sapmnt/QAS/profile/QAS_ASCS00_sapqasas", []byte{}, 0644)
```
works for us

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver_test.go:
```diff
@@ -0,0 +1,154 @@
+	factResults, err := g.Gather(factRequests)
+
+	suite.Nil(factResults)
+	suite.EqualError(err, "fact gathering error: saplocalhost_resolver-resolution-error - error resolving hostname: lookup sapqasas on 169.254.169.254:53: dial udp 169.254.169.254:53: connect: no route to host")
```
I find this outcome strange. Shouldn't we return an empty list of addresses here?
Otherwise, if we have 4 hostnames, and the resolution of one of them fails, the returned fact gathering would be failure, and maybe other hostnames resolve properly
PD: Commented in the prod code as well

arbulu89 commented on test/fixtures/gatherers/sap_profile.er.minimal:
```diff
@@ -0,0 +1,3 @@
+SAPSYSTEMNAME = NWP
```
I don't think we need this new file

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver_test.go:
```diff
@@ -0,0 +1,154 @@
+	err := appFS.MkdirAll("/usr/sap/QAS", 0644)
+	suite.NoError(err)
+
+	ascsProfileFile, _ := os.Open(helpers.GetFixturePath("gatherers/sap_profile.ascs"))
```
No need to get the file content

arbulu89 commented on internal/factsengine/gatherers/saplocalhostresolver_test.go:
```diff
@@ -0,0 +1,154 @@
+
+	erProfileFile, _ := os.Open(helpers.GetFixturePath("gatherers/sap_profile.er.minimal"))
+	erProfileConcent, _ := io.ReadAll(erProfileFile)
+	err = afero.WriteFile(appFS, "/sapmnt/QAS/profile/NWP_ER10_sapnwper", erProfileConcent, 0644)
```
This path is wrong. If should have `QAS` as `NWP` or viceversa. Not both. And better `ERS` than `ER`
By the way, I think we should have some test with a couple of SIDs and a couple of profiles

arbulu89 commented on internal/factsengine/gatherers/mocks/HostPinger.go:
```diff
@@ -0,0 +1,38 @@
+// Code generated by mockery v2.36.0. DO NOT EDIT.
```
Really strange, the 2 mocks are created with a different version of `mockery`

arbulu89 commented on go.mod:
```diff
@@ -14,28 +14,28 @@ require (
 	github.com/tklauser/go-sysconf v0.3.10 // indirect
 	github.com/trento-project/contracts/go v0.0.0-20230823130307-95ed2147fa9d
-	github.com/vektra/mockery/v2 v2.32.3
+	github.com/vektra/mockery/v2 v2.36.0
```
Do we need this update?
If so, please, use the same mockery version to create the mocks.
(I would abstein for updating for the sake of updating, otherwise, all of use now need to update)

arbulu89 commented on go.mod:
```diff
@@ -14,28 +14,28 @@ require (
 
 require (
-	github.com/chigopher/pathlib v1.0.0 // indirect
+	github.com/chigopher/pathlib v0.15.0 // indirect
```
Going back in version?
Strange. Please have a look on this and try to get a clean update of the file

rtorrero replied:
Missed this non-requested change, I guess from `go mod`, removed (as well as the mockery one)

arbulu89 reviewed (changes requested):
Suggestions from my side

rtorrero commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,177 @@
+	if err != nil {
+		return false
+	}
+	pinger.Count = 3
```
it could create false positives, but after a bit of thought it's actually something that we would want to capture, so I'll bring it down to 1

CDimonaco commented on internal/factsengine/gatherers/saplocalhostresolver.go:
```diff
@@ -0,0 +1,177 @@
+}
+
+func (r *SapLocalhostResolverGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	facts := make([]entities.Fact, 0)
```
Yes, this was my suggestion also, we don't need to preallocate anything

abravosuse commented:
@rtorrero @arbulu89 It's not the instance name that we are trying to resolve but the instance (virtual) hostname. Therefore my preference would be `SapInstanceHostnameResolver`.

rtorrero commented on internal/factsengine/gatherers/sapinstancehostnameresolver.go:
```diff
@@ -14,19 +14,19 @@ import (
+	hostnameRegexCompiled                   = regexp.MustCompile(`(.+)_(.+)_(.+)`) // <SID>_<InstanceNumber>_<Hostname>
+	regexSubgroupsCount                     = 4
+	SapInstanceHostnameResolverDetailsError = entities.FactGatheringError{
+		Type:    "sapinstance_hostname_resolver-details-error",
```
no worries, I can change it

EMaksy reviewed (commented):
Hey nice work, lgtm :D
